consensus/src/lib.rs

```diff
@@ -97,6 +100,55 @@ cfg_if::cfg_if! {
             ) -> impl Future<Output = oneshot::Receiver<bool>> + Send;
         }
 
+        /// Application is the interface responsible for building new blocks on top of consensus-provided parent
+        /// commitments as well as receiving finalized updates from [crate::marshal].
+        pub trait Application<E>: Clone + Send + 'static
+        where
+            E: Rng + Spawner + Metrics + Clock
+        {
+            /// Context is metadata provided by the consensus engine associated with a given payload.
+            ///
+            /// This often includes things like the proposer, view number, the height, or the epoch.
+            type Context: Epochable;
+
+            /// The block type produced by the application's builder.
+            type Block: Block;
+
+            /// Payload used to initialize the consensus engine in the first epoch.
+            fn genesis(&mut self) -> impl Future<Output = Self::Block> + Send;
+
+            /// Build a new block on top of the provided parent commitment. If the build job fails,
+            /// returns [None].
+            fn build(
+                &mut self,
+                r_context: E,
+                context: Self::Context,
+            ) -> impl Future<Output = Option<Self::Block>> + Send;
+
+            /// Receive a finalized block from [crate::marshal].
+            fn finalize(&mut self, block: Self::Block) -> impl Future<Output = ()> + Send;
+        }
+
+        /// An extension of [Application] that can verify blocks produced by its builder
+        /// for the sake of driving [Automaton::verify].
+        ///
+        /// Some [Application]s may not require this functionality - i.e. when employing
```

Comment:
I think we'll still verify during certification here tho? I was just thinking we'd pick when to ask the application to verify "auto-magically" (automatically handling chunks if coding) so that an application can activate coding ~unchanged.